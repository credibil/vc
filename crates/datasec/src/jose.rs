//! # JSON Object Signing and Encryption (JOSE) Proofs
//!
//! [JOSE] proofs are enveloping proofs for Credentials based on JWT [RFC7519],
//! JWS [RFC7515], and JWK [RFC7517].
//!
//! The Securing Verifiable Credentials using JOSE and COSE [VC-JOSE-COSE]
//! recommendation defines a "bridge" between these and the Verifiable
//! Credentials Data Model v2.0, specifying the suitable header claims, media
//! types, etc.
//!
//! In the case of JOSE, the Credential is the "payload". This is preceded by a
//! suitable header whose details are specified by Securing Verifiable
//! Credentials using JOSE and COSE for the usage of JWT. These are encoded,
//! concatenated, and signed, to be transferred in a compact form by one entity
//! to an other (e.g., sent by the holder to the verifier). All the intricate
//! details on signatures, encryption keys, etc., are defined by the IETF
//! specifications; see Example 6 for a specific case.
//!
//! ## Note
//!
//! If the JWT is only a JWE, iss, exp and aud MUST be omitted in the JWT Claims
//! Set of the JWE, and the processing rules as per JARM Section 2.4 related to
//! these claims do not apply. [OpenID4VP] JWT - JWE
//!
//! ```json
//! {
//!   "vp_token": "eyJI...",
//!   "presentation_submission": {...}
//! }
//! ```
//!
//! [JOSE]: https://datatracker.ietf.org/wg/jose/about
//! [RFC7515]: https://www.rfc-editor.org/rfc/rfc7515
//! [RFC7517]: https://www.rfc-editor.org/rfc/rfc7517
//! [RFC7519]: https://www.rfc-editor.org/rfc/rfc7519
//! [VC-JOSE-COSE]: https://w3c.github.io/vc-jose-cose
//! [OpenID4VP]: https://openid.net/specs/openid-4-verifiable-presentations-1_0.html

pub mod jwa;
pub mod jwe;
pub mod jwk;
pub mod jws;
pub mod jwt;
