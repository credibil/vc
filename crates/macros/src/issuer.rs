use proc_macro2::{Span, TokenStream};
use quote::quote;
use syn::parse::{Error, Result};

use crate::parse::Data;

pub fn create_offer_request(input: &Data) -> Result<TokenStream> {
    // remove fields so we can error on any remaining
    let mut fields = input.fields.clone();
    let credential_issuer = fields.remove("credential_issuer").unwrap();
    let subject_id = fields.remove("subject_id").unwrap();
    let credential_configuration_ids = fields.remove("credential_configuration_ids").unwrap();
    let pre_authorize = fields.remove("pre-authorize").unwrap();
    let tx_code_required = fields.remove("tx_code_required").unwrap();
    let send_type = fields.remove("send_type").unwrap();

    // return error for any additional fields
    if !fields.is_empty() {
        let keys = fields.keys().map(|k| format!("`{k}`")).collect::<Vec<_>>().join(", ");
        return Err(Error::new(Span::call_site(), format!("unexpected field(s): {keys}")));
    }

    Ok(quote! {
        CreateOfferRequest {
            credential_issuer: #credential_issuer.to_string(),
            subject_id: Some(#subject_id.to_string()),
            credential_configuration_ids: #credential_configuration_ids.iter().map(|s| s.to_string()).collect(),
            pre_authorize: #pre_authorize,
            tx_code_required: #tx_code_required,
            send_type:  #send_type,
        }
    })
}

pub fn authorization_request(input: &Data) -> Result<TokenStream> {
    // remove fields so we can error on any remaining
    let mut fields = input.fields.clone();
    let credential_issuer = fields.remove("credential_issuer").unwrap();
    let response_type = fields.remove("response_type").unwrap();
    let client_id = fields.remove("client_id").unwrap();
    let redirect_uri = fields.remove("redirect_uri").unwrap();
    let state = fields.remove("state").unwrap();
    //  let code_challenge = fields.remove("code_challenge").unwrap();
    let code_challenge_method = fields.remove("code_challenge_method").unwrap();

    // return error for any additional fields
    if !fields.is_empty() {
        let keys = fields.keys().map(|k| format!("`{k}`")).collect::<Vec<_>>().join(", ");
        return Err(Error::new(Span::call_site(), format!("unexpected field(s): {keys}")));
    }

    Ok(quote! {
        AuthorizationRequest {
            credential_issuer: #credential_issuer.to_string(),
            response_type: #response_type.to_string(),
            client_id: #client_id.to_string(),
            redirect_uri: Some(#redirect_uri.to_string()),
            state: Some(#state.to_string()),
            // code_challenge: #code_challenge.to_string(),
            code_challenge_method: #code_challenge_method.to_string(),

            // credential_configuration_ids: #credential_configuration_ids.iter().map(|s| s.to_string()).collect(),

            ..Default::default()
        }
    })
}
