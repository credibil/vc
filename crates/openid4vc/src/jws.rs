//! # JOSE Proofs
//!
//! JSON Object Signing and Encryption ([JOSE]) proofs are a form of enveloping proofs
//! of Credentials based on JWT [RFC7519], JWS [RFC7515], and JWK [RFC7517].
//!
//! The Securing Verifiable Credentials using JOSE and COSE [VC-JOSE-COSE]
//! recommendation defines a "bridge" between these and the Verifiable Credentials Data
//! Model v2.0, specifying the suitable header claims, media types, etc.
//!
//! In the case of JOSE, the Credential is the "payload". This is preceded by a suitable
//! header whose details are specified by Securing Verifiable Credentials using JOSE and
//! COSE for the usage of JWT. These are encoded, concatenated, and signed, to be
//! transferred in a compact form by one entity to an other (e.g., sent by the holder to
//! the verifier). All the intricate details on signatures, encryption keys, etc., are
//! defined by the IETF specifications; see Example 6 for a specific case.
//!
//! ## Note
//!
//! If the JWT is only a JWE, iss, exp and aud MUST be omitted in the JWT Claims
//! Set of the JWE, and the processing rules as per JARM Section 2.4 related to
//! these claims do not apply. [OpenID4VP] JWT - JWE
//!
//! ```json
//! {
//!   "vp_token": "eyJI...",
//!   "presentation_submission": {...}
//! }
//! ```
//!
//! [JOSE]: https://datatracker.ietf.org/wg/jose/about
//! [RFC7519]: https://www.rfc-editor.org/rfc/rfc7519
//! [RFC7517]: https://www.rfc-editor.org/rfc/rfc7517
//! [VC-JOSE-COSE]: https://w3c.github.io/vc-jose-cose
//! [OpenID4VP]: https://openid.net/specs/openid-4-verifiable-presentations-1_0.html

use std::fmt::{Debug, Display};

use anyhow::{anyhow, bail};
use base64ct::{Base64UrlUnpadded, Encoding};
use ecdsa::signature::Verifier as _;
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};

use crate::endpoint::{Algorithm, Jwk, Signer, Verifier};

/// The JWT `typ` claim.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub enum Type {
    /// JWT `typ` for Verifiable Credential.
    #[default]
    #[serde(rename = "jwt")]
    Credential,

    /// JWT `typ` for Verifiable Presentation.
    #[serde(rename = "jwt")]
    Presentation,

    /// JWT `typ` for Authorization Request Object.
    #[serde(rename = "oauth-authz-req+jwt")]
    Request,

    /// JWT `typ` for Wallet's Proof of possession of key material.
    #[serde(rename = "openid4vci-proof+jwt")]
    Proof,
}

impl Display for Type {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{self:?}")
    }
}

/// Encode the provided header and claims and sign, returning a JWT in compact JWS form.
///
/// # Errors
/// TODO: add error docs
pub async fn encode<T>(typ: Type, claims: &T, signer: impl Signer) -> anyhow::Result<String>
where
    T: Serialize + Send + Sync,
{
    tracing::debug!("encode");

    // header
    let header = Header {
        alg: signer.algorithm(),
        typ,
        kid: Some(signer.verification_method()),
        ..Header::default()
    };

    // payload
    let header_raw = serde_json::to_vec(&header)?;
    let header_enc = Base64UrlUnpadded::encode_string(&header_raw);
    let claims_raw = serde_json::to_vec(claims)?;
    let claims_enc = Base64UrlUnpadded::encode_string(&claims_raw);
    let payload = format!("{header_enc}.{claims_enc}");

    // sign
    let sig = signer.try_sign(payload.as_bytes()).await?;
    let sig_enc = Base64UrlUnpadded::encode_string(&sig);

    Ok(format!("{payload}.{sig_enc}"))
}

// TODO: allow passing verifier into this method
/// Decode the JWT token and return the claims.
///
/// # Errors
/// TODO: Add errors
pub async fn decode<T>(token: &str, verifier: &impl Verifier) -> anyhow::Result<Jwt<T>>
where
    T: DeserializeOwned + Send,
{
    // TODO: cater for different key types
    let parts = token.split('.').collect::<Vec<&str>>();
    if parts.len() != 3 {
        bail!("invalid Compact JWS format");
    }

    // deserialize header, claims, and signature
    let decoded = Base64UrlUnpadded::decode_vec(parts[0])
        .map_err(|e| anyhow!("issue decoding header: {e}"))?;
    let header: Header =
        serde_json::from_slice(&decoded).map_err(|e| anyhow!("issue deserializing header: {e}"))?;
    let decoded = Base64UrlUnpadded::decode_vec(parts[1])
        .map_err(|e| anyhow!("issue decoding claims: {e}"))?;
    let claims =
        serde_json::from_slice(&decoded).map_err(|e| anyhow!("issue deserializing claims:{e}"))?;
    let sig = Base64UrlUnpadded::decode_vec(parts[2])
        .map_err(|e| anyhow!("issue decoding signature: {e}"))?;

    // check algorithm
    if !(header.alg == Algorithm::ES256K || header.alg == Algorithm::EdDSA) {
        bail!("'alg' is not recognised");
    }

    // verify signature
    let jwk = verifier.deref_jwk(&header.kid.clone().unwrap_or_default()).await?;
    verify(&jwk, &format!("{}.{}", parts[0], parts[1]), &sig)?;

    Ok(Jwt { header, claims })
}

/// Verify the signature of the provided message using the JWK.
///
/// # Errors
///
/// Will return an error if the signature is invalid, the JWK is invalid, or the
/// algorithm is unsupported.
pub fn verify(jwk: &Jwk, msg: &str, sig: &[u8]) -> anyhow::Result<()> {
    match jwk.crv.as_str() {
        "ES256K" | "secp256k1" => verify_es256k(jwk, msg, sig), // kty: "EC"
        "X25519" => verify_eddsa(jwk, msg, sig),                // kty: "OKP"
        _ => bail!("Unsupported JWT signature algorithm"),
    }
}

// Verify the signature of the provided message using the ES256K algorithm.
fn verify_es256k(jwk: &Jwk, msg: &str, sig: &[u8]) -> anyhow::Result<()> {
    use ecdsa::{Signature, VerifyingKey};
    use k256::Secp256k1;

    // build verifying key
    let y = jwk.y.as_ref().ok_or_else(|| anyhow!("Proof JWT 'y' is invalid"))?;
    let mut sec1 = vec![0x04]; // uncompressed format
    sec1.append(&mut Base64UrlUnpadded::decode_vec(&jwk.x)?);
    sec1.append(&mut Base64UrlUnpadded::decode_vec(y)?);

    let verifying_key = VerifyingKey::<Secp256k1>::from_sec1_bytes(&sec1)?;
    let signature: Signature<Secp256k1> = Signature::from_slice(sig)?;

    Ok(verifying_key.verify(msg.as_bytes(), &signature)?)
}

// Verify the signature of the provided message using the EdDSA algorithm.
fn verify_eddsa(jwk: &Jwk, msg: &str, sig_bytes: &[u8]) -> anyhow::Result<()> {
    use ed25519_dalek::{Signature, VerifyingKey};

    // build verifying key
    let x_bytes = Base64UrlUnpadded::decode_vec(&jwk.x)
        .map_err(|e| anyhow!("unable to base64 decode proof JWK 'x': {e}"))?;
    let bytes = &x_bytes.try_into().map_err(|_| anyhow!("invalid public key length"))?;
    let verifying_key = VerifyingKey::from_bytes(bytes)
        .map_err(|e| anyhow!("unable to build verifying key: {e}"))?;
    let signature =
        Signature::from_slice(sig_bytes).map_err(|e| anyhow!("unable to build signature: {e}"))?;

    verifying_key
        .verify(msg.as_bytes(), &signature)
        .map_err(|e| anyhow!("unable to verify signature: {e}"))
}

/// Represents a JWT as used for proof and credential presentation.
#[derive(Clone, Debug, Default, Serialize, PartialEq, Eq)]
pub struct Jwt<T> {
    /// The JWT header.
    pub header: Header,

    /// The JWT claims.
    pub claims: T,
}

/// Represents the JWT header.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub struct Header {
    /// Digital signature algorithm identifier as per IANA "JSON Web Signature
    /// and Encryption Algorithms" registry.
    pub alg: Algorithm,

    /// Used to declare the media type [IANA.MediaTypes](http://www.iana.org/assignments/media-types)
    /// of the JWS.
    pub typ: Type,

    /// Contains the key ID. If the Credential is bound to a DID, the kid refers to a
    /// DID URL which identifies a particular key in the DID Document that the
    /// Credential should bound to. Alternatively, may refer to  a key inside a JWKS.
    ///
    /// MUST NOT be set if `jwk` property is set.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub kid: Option<String>,

    /// Contains the key material the new Credential shall be bound to.
    ///
    /// MUST NOT be set if `kid` is set.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub jwk: Option<Jwk>,

    /// Contains a certificate (or certificate chain) corresponding to the key used to
    /// sign the JWT. This element MAY be used to convey a key attestation. In such a
    /// case, the actual key certificate will contain attributes related to the key
    /// properties.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub x5c: Option<String>,

    /// Contains an OpenID.Federation Trust Chain. This element MAY be used to convey
    /// key attestation, metadata, metadata policies, federation Trust Marks and any
    /// other information related to a specific federation, if available in the chain.
    ///
    /// When used for signature verification, `kid` MUST be set.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trust_chain: Option<String>,
}
