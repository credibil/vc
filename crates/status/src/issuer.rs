//! # Status (Issuer)
//!
//! Traits and types for managing the publication of a credential status as an
//! issuer.

use std::future::Future;

pub use vercre_core::Quota;
pub use vercre_w3c_vc::model::CredentialStatus;

use crate::provider;

/// The `Status` trait specifies how status information can be looked up for a
/// credential.
pub trait Status: Send + Sync {
    /// Returns information on how to look up the status of a credential.
    ///
    /// A default implementation is provided that just returns `None` for cases
    /// where the issuer will not be providing a status endpoint.
    fn status(
        &self, _holder_subject: &str, _credential_identifier: &str,
    ) -> impl Future<Output = provider::Result<Option<Quota<CredentialStatus>>>> + Send {
        async { Ok(None) }
    }
}
