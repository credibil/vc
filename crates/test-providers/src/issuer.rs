use anyhow::Result;
use credibil_core::blockstore::BlockStore;
use credibil_identity::{Identity, IdentityResolver, Key, SignerExt};
use credibil_se::{Algorithm, Signer};

use crate::blockstore::Mockstore;
use crate::identity::DidIdentity;

pub const ISSUER_ID: &str = "http://credibil.io";
pub const BOB_ID: &str = "bob";
pub const CAROL_ID: &str = "carol";

pub mod data {
    pub const CLIENT: &[u8] = include_bytes!("../data/issuer/client.json");
    pub const ISSUER: &[u8] = include_bytes!("../data/issuer/issuer.json");
    pub const SERVER: &[u8] = include_bytes!("../data/issuer/server.json");
    pub const NORMAL_USER: &[u8] = include_bytes!("../data/issuer/normal-user.json");
    pub const PENDING_USER: &[u8] = include_bytes!("../data/issuer/pending-user.json");
}

#[derive(Clone)]
pub struct Issuer {
    identity: DidIdentity,
    blockstore: Mockstore,
}

impl Issuer {
    #[must_use]
    pub async fn new(owner: &str) -> Self {
        Self {
            identity: DidIdentity::new(owner).await,
            blockstore: Mockstore::new(),
        }
    }
}

impl IdentityResolver for Issuer {
    async fn resolve(&self, url: &str) -> anyhow::Result<Identity> {
        self.identity.resolve(url).await
    }
}

impl Signer for Issuer {
    async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        self.identity.try_sign(msg).await
    }

    async fn verifying_key(&self) -> Result<Vec<u8>> {
        self.identity.verifying_key().await
    }

    async fn algorithm(&self) -> Result<Algorithm> {
        Ok(self.identity.algorithm())
    }
}

impl SignerExt for Issuer {
    async fn verification_method(&self) -> Result<Key> {
        self.identity.verification_method().await
    }
}

impl BlockStore for Issuer {
    async fn put(&self, owner: &str, partition: &str, key: &str, block: &[u8]) -> Result<()> {
        self.blockstore.put(owner, partition, key, block).await
    }

    async fn get(&self, owner: &str, partition: &str, key: &str) -> Result<Option<Vec<u8>>> {
        self.blockstore.get(owner, partition, key).await
    }

    async fn delete(&self, owner: &str, partition: &str, key: &str) -> Result<()> {
        self.blockstore.delete(owner, partition, key).await
    }

    async fn purge(&self, _owner: &str, _partition: &str) -> Result<()> {
        unimplemented!()
    }
}
