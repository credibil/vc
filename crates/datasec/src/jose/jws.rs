//! # JSON Web Signature (JWS)
//!
//! JWS ([RFC7515]) represents content secured with digital signatures using JSON-based data
//! structures. Cryptographic algorithms and identifiers for use with this
//! specification are described in the JWA ([RFC7518]) specification.
//!
//! [RFC7515]: https://www.rfc-editor.org/rfc/rfc7515
//! [RFC7518]: https://www.rfc-editor.org/rfc/rfc7518

use anyhow::{anyhow, bail};
use base64ct::{Base64UrlUnpadded, Encoding};
use ecdsa::signature::Verifier as _;
use futures::Future;
use serde::de::DeserializeOwned;
use serde::Serialize;

use crate::jose::jwk::{Curve, PublicKeyJwk};
pub use crate::jose::jwt::{Header, Jwt, KeyType, Type};
use crate::{Algorithm, Signer};

/// Encode the provided header and claims and sign, returning a JWT in compact JWS form.
///
/// # Errors
/// TODO: add error docs
pub async fn encode<T>(typ: Type, claims: &T, signer: impl Signer) -> anyhow::Result<String>
where
    T: Serialize + Send + Sync,
{
    tracing::debug!("encode");

    // header
    let header = Header {
        alg: signer.algorithm(),
        typ,
        key: KeyType::KeyId(signer.verification_method()),
        ..Header::default()
    };

    // payload
    let header = Base64UrlUnpadded::encode_string(&serde_json::to_vec(&header)?);
    let claims = Base64UrlUnpadded::encode_string(&serde_json::to_vec(claims)?);
    let payload = format!("{header}.{claims}");

    // sign
    let sig = signer.try_sign(payload.as_bytes()).await?;
    let sig_enc = Base64UrlUnpadded::encode_string(&sig);

    Ok(format!("{payload}.{sig_enc}"))
}

// TODO: allow passing verifier into this method
/// Decode the JWT token and return the claims.
///
/// # Errors
/// TODO: Add errors
pub async fn decode<F, Fut, T>(token: &str, pk_cb: F) -> anyhow::Result<Jwt<T>>
where
    T: DeserializeOwned + Send,
    F: FnOnce(String) -> Fut + Send + Sync,
    Fut: Future<Output = anyhow::Result<PublicKeyJwk>> + Send + Sync,
{
    // TODO: cater for different key types
    let parts = token.split('.').collect::<Vec<&str>>();
    if parts.len() != 3 {
        bail!("invalid Compact JWS format");
    }

    // deserialize header, claims, and signature
    let decoded = Base64UrlUnpadded::decode_vec(parts[0])
        .map_err(|e| anyhow!("issue decoding header: {e}"))?;
    let header: Header =
        serde_json::from_slice(&decoded).map_err(|e| anyhow!("issue deserializing header: {e}"))?;
    let decoded = Base64UrlUnpadded::decode_vec(parts[1])
        .map_err(|e| anyhow!("issue decoding claims: {e}"))?;
    let claims =
        serde_json::from_slice(&decoded).map_err(|e| anyhow!("issue deserializing claims:{e}"))?;
    let sig = Base64UrlUnpadded::decode_vec(parts[2])
        .map_err(|e| anyhow!("issue decoding signature: {e}"))?;

    // check algorithm
    if !(header.alg == Algorithm::ES256K || header.alg == Algorithm::EdDSA) {
        bail!("'alg' is not recognised");
    }

    // verify signature
    let KeyType::KeyId(kid) = header.key.clone() else {
        bail!("'kid' is not set");
    };

    // resolve 'kid' to Jwk (hint: kid will contain a DID URL for now)
    let jwk = pk_cb(kid).await?;
    verify(&jwk, &format!("{}.{}", parts[0], parts[1]), &sig)?;

    Ok(Jwt { header, claims })
}

/// Verify the signature of the provided message using the JWK.
///
/// # Errors
///
/// Will return an error if the signature is invalid, the JWK is invalid, or the
/// algorithm is unsupported.
pub fn verify(jwk: &PublicKeyJwk, msg: &str, sig: &[u8]) -> anyhow::Result<()> {
    match jwk.crv {
        Curve::Es256K => verify_es256k(jwk, msg, sig),
        Curve::Ed25519 => verify_eddsa(jwk, msg, sig),
    }
}

// Verify the signature of the provided message using the ES256K algorithm.
fn verify_es256k(jwk: &PublicKeyJwk, msg: &str, sig: &[u8]) -> anyhow::Result<()> {
    use ecdsa::{Signature, VerifyingKey};
    use k256::Secp256k1;

    // build verifying key
    let y = jwk.y.as_ref().ok_or_else(|| anyhow!("Proof JWT 'y' is invalid"))?;
    let mut sec1 = vec![0x04]; // uncompressed format
    sec1.append(&mut Base64UrlUnpadded::decode_vec(&jwk.x)?);
    sec1.append(&mut Base64UrlUnpadded::decode_vec(y)?);

    let verifying_key = VerifyingKey::<Secp256k1>::from_sec1_bytes(&sec1)?;
    let signature: Signature<Secp256k1> = Signature::from_slice(sig)?;
    let normalised = signature.normalize_s().unwrap_or(signature);

    Ok(verifying_key.verify(msg.as_bytes(), &normalised)?)
}

// Verify the signature of the provided message using the EdDSA algorithm.
fn verify_eddsa(jwk: &PublicKeyJwk, msg: &str, sig_bytes: &[u8]) -> anyhow::Result<()> {
    use ed25519_dalek::{Signature, VerifyingKey};

    // build verifying key
    let x_bytes = Base64UrlUnpadded::decode_vec(&jwk.x)
        .map_err(|e| anyhow!("unable to base64 decode proof JWK 'x': {e}"))?;
    let bytes = &x_bytes.try_into().map_err(|_| anyhow!("invalid public key length"))?;
    let verifying_key = VerifyingKey::from_bytes(bytes)
        .map_err(|e| anyhow!("unable to build verifying key: {e}"))?;
    let signature =
        Signature::from_slice(sig_bytes).map_err(|e| anyhow!("unable to build signature: {e}"))?;

    verifying_key
        .verify(msg.as_bytes(), &signature)
        .map_err(|e| anyhow!("unable to verify signature: {e}"))
}

// /// In the JWS JSON Serialization, one or both of the JWS Protected Header and
// /// JWS Unprotected Header MUST be present.  In this case, the members of the
// /// JOSE Header are the union of the members of the JWS Protected Header and
// /// the JWS Unprotected Header values that are present.
// #[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
// pub struct Jwe {
//     /// JWE protected header.
//     protected: Header,

//     /// Shared unprotected header as a JSON object.
//     #[serde(skip_serializing_if = "Option::is_none")]
//     unprotected: Option<Value>,

//     /// Encrypted key, as a base64Url encoded string.
//     payload: String,

//     /// AAD value, base64url encoded. Not used for JWE Compact Serialization.
//     signature: String,
// }
