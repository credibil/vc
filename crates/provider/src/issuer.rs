#![allow(unused)]

use anyhow::Result;
use credibil_did::{DidResolver, Document};
use credibil_infosec::{Algorithm, Signer};
use credibil_vc::BlockStore;
use credibil_vc::status::issuer::Status;

use crate::blockstore::Mockstore;
use crate::keystore::Keyring;

pub const ISSUER_ID: &str = "http://credibil.io";
pub const BOB_ID: &str = "bob";
pub const CAROL_ID: &str = "carol";

pub mod data {
    pub const CLIENT: &[u8] = include_bytes!("../data/issuer/client.json");
    pub const ISSUER: &[u8] = include_bytes!("../data/issuer/issuer.json");
    pub const SERVER: &[u8] = include_bytes!("../data/issuer/server.json");
    pub const NORMAL_USER: &[u8] = include_bytes!("../data/issuer/normal-user.json");
    pub const PENDING_USER: &[u8] = include_bytes!("../data/issuer/pending-user.json");
}

#[derive(Clone, Debug)]
pub struct Issuer {
    keyring: Keyring,
    blockstore: Mockstore,
}

impl Issuer {
    #[must_use]
    pub fn new() -> Self {
        Self {
            keyring: Keyring::new(),
            blockstore: Mockstore::new(),
        }
    }
}

impl DidResolver for Issuer {
    async fn resolve(&self, url: &str) -> anyhow::Result<Document> {
        self.keyring.resolve(url).await
    }
}

impl Signer for Issuer {
    async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        self.keyring.try_sign(msg).await
    }

    async fn verifying_key(&self) -> Result<Vec<u8>> {
        self.keyring.verifying_key().await
    }

    fn algorithm(&self) -> Algorithm {
        self.keyring.algorithm()
    }

    async fn verification_method(&self) -> Result<String> {
        self.keyring.verification_method().await
    }
}

impl BlockStore for Issuer {
    async fn put(&self, owner: &str, partition: &str, key: &str, block: &[u8]) -> Result<()> {
        self.blockstore.put(owner, partition, key, block).await
    }

    async fn get(&self, owner: &str, partition: &str, key: &str) -> Result<Option<Vec<u8>>> {
        self.blockstore.get(owner, partition, key).await
    }

    async fn delete(&self, owner: &str, partition: &str, key: &str) -> Result<()> {
        self.blockstore.delete(owner, partition, key).await
    }

    async fn purge(&self, _owner: &str, _partition: &str) -> Result<()> {
        unimplemented!()
    }
}

impl Status for Issuer {}
