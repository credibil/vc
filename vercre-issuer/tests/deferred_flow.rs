use std::sync::LazyLock;

use assert_let_bind::assert_let;
use chrono::Utc;
use datasec::jose::jws::{self, Type};
use futures::future::TryFutureExt;
use insta::assert_yaml_snapshot as assert_snapshot;
use serde_json::json;
use test_utils::holder;
use test_utils::issuer::{self, CLIENT_ID, CREDENTIAL_ISSUER, PENDING_USER};
use vercre_issuer::{
    CreateOfferRequest, CreateOfferResponse, CredentialOfferType, CredentialRequest,
    CredentialResponse, DeferredCredentialRequest, DeferredCredentialResponse, ProofClaims,
    TokenRequest, TokenResponse,
};

static ISSUER_PROVIDER: LazyLock<issuer::Provider> = LazyLock::new(issuer::Provider::new);

// Run through entire pre-authorized code flow.
#[tokio::test]
async fn deferred_flow() {
    test_utils::init_tracer();

    // go through the pre-auth flow to token endpoint
    let token_resp = get_offer().and_then(get_token).await.expect("Ok");
    let credential_resp = get_credential(token_resp.clone()).await.expect("Ok");

    assert_snapshot!("credential-response", &credential_resp, {
        ".transaction_id" => "[transaction_id]",
        ".c_nonce" => "[c_nonce]",
        ".c_nonce_expires_in" => "[c_nonce_expires_in]"
    });

    let deferred_resp = get_deferred(token_resp, credential_resp.clone()).await.expect("Ok");

    assert_snapshot!("deferred-response", deferred_resp, {
        ".credential" => "[credential]",
        ".c_nonce" => "[c_nonce]",
        ".c_nonce_expires_in" => "[c_nonce_expires_in]"
    });
}

// Simulate Issuer request to '/create_offer' endpoint to get credential offer to use to
// make credential offer to Wallet.
async fn get_offer() -> openid::Result<CreateOfferResponse> {
    // offer request
    let body = json!({
        "credential_configuration_ids": ["EmployeeID_JWT"],
        "subject_id": PENDING_USER,
        "pre-authorize": true,
        "tx_code_required": true,
        "callback_id": "1234"
    });

    let mut request =
        serde_json::from_value::<CreateOfferRequest>(body).expect("should deserialize");
    request.credential_issuer = CREDENTIAL_ISSUER.into();

    vercre_issuer::create_offer(ISSUER_PROVIDER.clone(), &request).await
}

// Simulate Wallet request to '/token' endpoint with pre-authorized code to get
// access token
async fn get_token(input: CreateOfferResponse) -> openid::Result<TokenResponse> {
    assert_let!(CredentialOfferType::Object(offer), &input.credential_offer);

    assert_let!(Some(grants), &offer.grants);
    assert_let!(Some(pre_authorized_code), &grants.pre_authorized_code);

    // create TokenRequest to 'send' to the app
    let body = json!({
        "client_id": CLIENT_ID,
        "grant_type": "urn:ietf:params:oauth:grant-type:pre-authorized_code",
        "pre-authorized_code": &pre_authorized_code.pre_authorized_code,
        "user_code": input.user_code.unwrap_or_default(),
    });

    let mut request = serde_json::from_value::<TokenRequest>(body).expect("should deserialize");
    request.credential_issuer = CREDENTIAL_ISSUER.into();

    vercre_issuer::token(ISSUER_PROVIDER.clone(), &request).await
}

// Simulate Wallet request to '/credential' endpoint with access token to get credential.
async fn get_credential(input: TokenResponse) -> openid::Result<CredentialResponse> {
    // create CredentialRequest to 'send' to the app
    let claims = ProofClaims {
        iss: Some(CLIENT_ID.into()),
        aud: CREDENTIAL_ISSUER.into(),
        iat: Utc::now().timestamp(),
        nonce: input.c_nonce,
    };
    let jwt = jws::encode(Type::Proof, &claims, holder::Provider).await.expect("should encode");

    let body = json!({
        "format": "jwt_vc_json",
        "credential_definition": {
            "type": [
                "VerifiableCredential",
                "EmployeeIDCredential"
            ]
        },
        "proof":{
            "proof_type": "jwt",
            "jwt": jwt
        }
    });

    let mut request =
        serde_json::from_value::<CredentialRequest>(body).expect("should deserialize");
    request.credential_issuer = CREDENTIAL_ISSUER.into();
    request.access_token = input.access_token;

    vercre_issuer::credential(ISSUER_PROVIDER.clone(), &request).await
}

async fn get_deferred(
    tkn: TokenResponse, cred_resp: CredentialResponse,
) -> openid::Result<DeferredCredentialResponse> {
    let request = DeferredCredentialRequest {
        credential_issuer: CREDENTIAL_ISSUER.into(),
        access_token: tkn.access_token,
        transaction_id: cred_resp.transaction_id.expect("should have transaction_id"),
    };

    vercre_issuer::deferred(ISSUER_PROVIDER.clone(), &request).await
}
