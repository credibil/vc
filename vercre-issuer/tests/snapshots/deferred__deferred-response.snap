---
source: vercre-issuer/tests/deferred_flow.rs
assertion_line: 39
expression: deferred_resp
---
credential: "[credential]"
c_nonce: "[c_nonce]"
c_nonce_expires_in: "[c_nonce_expires_in]"
